Packages/com.unity.inputsystem/InputSystem/Plugins/UI/TrackedDeviceRaycaster.cs

```diff
@@ -1,12 +1,23 @@
-#if UNITY_INPUT_SYSTEM_ENABLE_XR
+#if UNITY_INPUT_SYSTEM_ENABLE_XR || UNITY_EDITOR || PACKAGE_DOCS_GENERATION
 using System;
 using System.Collections.Generic;
 using UnityEngine.EventSystems;
+using UnityEngine.InputSystem.Utilities;
+using UnityEngine.Serialization;
 using UnityEngine.UI;
 
-#pragma warning disable CS0649
 namespace UnityEngine.InputSystem.UI
 {
+    /// <summary>
+    /// Racyasting implementation for use with <see cref="TrackedDevice"/>s.
+    /// </summary>
+    /// <remarks>
+    /// This component needs to be added alongside the <c>Canvas</c> component. Usually, raycasting is
+    /// performed by the <c>GraphicRaycaster</c> component found there but for 3D raycasting necessary for
+    /// tracked devices, this component is required.
+    /// </remarks>
+    [AddComponentMenu("Event/Tracked Device Raycaster")]
+    [RequireComponent(typeof(Canvas))]
```

Comment:
Why do we require the canvas? I think it would be nice if this just worked for the whole EventSystem instead of just with UI.

Reply:
It seems the TrackedDeviceRaycaster needs a canvas to work. I'm not very familiar with the internals of this thing. @StayTalm wrote this raycaster. With the attributes here I mainly followed what `GraphicRaycaster` is doing. It also does the `RequirementComponent(typeof(Canvas))` thing.

Reply:
Yeah would have been interesting to extend it past just the UI but it does 1 get for the canvas elements so i know why its necessary. Maybe future improvement if we'd want.

Reply:
Would great if we can do this as part of the UI Elements work. My guess is that when it comes to implementing that, there's several things in the current uGUI support code of the input system that make sense to reuse. Hopefully we can generalize these at that point. Ideally, I'd love for users to be able to rather easily hook up their own UI systems, too. Maybe a pipe dream :)